Add permission policy checks

aykutbulut opened the pull request:
Add permission policy feature definitions. 

Update the Request constructor with policy checks in Modifications to request section.


<!--
    This comment and the below content is programmatically generated.
    You may add a comma-separated list of anchors you'd like a
    direct link to below (e.g. #idl-serializers, #idl-sequence):

    Don't remove this comment or modify anything below this line.
    If you don't want a preview generated for this pull request,
    just replace the whole of this comment's content by "no preview"
    and remove what's below.
-->
***
<a href="https://pr-preview.s3.amazonaws.com/aykutbulut/trust-token-api/pull/239.html" title="Last updated on May 8, 2023, 6:12 PM UTC (9916391)">Preview</a> | <a href="https://pr-preview.s3.amazonaws.com/WICG/trust-token-api/239/9927f3c...aykutbulut:9916391.html" title="Last updated on May 8, 2023, 6:12 PM UTC (9916391)">Diff</a>

aykutbulut commented:
I get the same error (copied below for convenience) with my local bikeshed installation. I am not sure why it can not find 
reference for "should request be allowed to use feature". It is defined in the following section https://www.w3.org/TR/permissions-policy/#algo-should-request-be-allowed-to-use-feature

```
LINE ~455: No 'dfn' refs found for 'should request be allowed to use feature'.
[=should request be allowed to use feature=]
```

jyasskin commented:
It's got `data-dfn-type="abstract-op"`, and it ends with a "?", so you have to use `[$Should request be allowed to use feature?$]`. Sorry.

aykutbulut commented:
Thanks for the tip. Fixed the reference. Added throwing NotAllowedError Dom exception before aborting the constructor steps.

Tried to add a custom text for the algorithm link and failed. Any ideas on how to fix that part?

jyasskin commented on spec.bs:
```diff
@@ -439,13 +439,21 @@ A [=/request=] has an associated <dfn for=request>private token issuers</dfn>, w
 
+This specification defines two new [=policy-controlled features=]. Exactly one of these policy features applies for a given Private State Token operation.
+
+Policy identified by "<dfn data-dfn-for="policy-controlled feature"><code>private-state-token-issuance</code></dfn>" applies for the <code>"token-request"</code> operation. The [=default allowlist=] for this feature is <code>["self"]</code>.
```
```suggestion
The [=policy-controlled feature=] identified by "<dfn data-dfn-for="policy-controlled feature"><code>private-state-token-issuance</code></dfn>" applies for the <code>"token-request"</code> operation. The [=default allowlist=] for this feature is <code>["self"]</code>.
```

jyasskin commented on spec.bs:
```diff
@@ -439,13 +439,21 @@ A [=/request=] has an associated <dfn for=request>private token issuers</dfn>, w
+
+Policy identified by "<dfn data-dfn-for="policy-controlled feature"><code>private-state-token-issuance</code></dfn>" applies for the <code>"token-request"</code> operation. The [=default allowlist=] for this feature is <code>["self"]</code>.
+
+Policy identified by "<dfn data-dfn-for="policy-controlled feature"><code>private-state-token-redemption</code></dfn>" applies for the <code>"send-redemption-record"</code> and <code>"token-redemption"</code> operations. The [=default allowlist=] for this feature is <code>["self"]</code>.
```
```suggestion
The [=policy-controlled feature=] identified by "<dfn data-dfn-for="policy-controlled feature"><code>private-state-token-redemption</code></dfn>" applies for the <code>"send-redemption-record"</code> and <code>"token-redemption"</code> operations. The [=default allowlist=] for this feature is <code>["self"]</code>.
```

jyasskin commented on spec.bs:
```diff
@@ -439,13 +439,21 @@ A [=/request=] has an associated <dfn for=request>private token issuers</dfn>, w
-     1. If |init|["{{RequestInit/privateToken}}"]["{{PrivateToken/operation}}"] is <code>"token-request"</code>, then abort these steps.
+     1. If |init|["{{RequestInit/privateToken}}"]["{{PrivateToken/operation}}"] is <code>"token-request"</code>:
+         1. If [$Should request be allowed to use feature?$] for "<code>[=policy-controlled feature/private-state-token-issuance=]</code>" returns false, then throw  "{{NotAllowedError}}" {{DOMException}} and abort the remaining steps.
+     1. If [$Should request be allowed to use feature?$] for "<code>[=policy-controlled feature/private-state-token-redemption=]</code>" returns false, then throw  "{{NotAllowedError}}" {{DOMException}} and abort the remaining steps.
```
This also runs for "token-request" operations. You should add a check for the other operation values.

aykutbulut replied:
Yes it is. Updated the flow a bit to fix this.

aykutbulut commented on spec.bs:
```diff
@@ -439,13 +439,21 @@ A [=/request=] has an associated <dfn for=request>private token issuers</dfn>, w
      1. Set <var ignore>request</var>'s [=request/private token operation=]</a> to |init|["{{RequestInit/privateToken}}"]["{{PrivateToken/operation}}"].
-     1. If |init|["{{RequestInit/privateToken}}"]["{{PrivateToken/operation}}"] is <code>"token-request"</code>, then abort these steps.
+     1. If |init|["{{RequestInit/privateToken}}"]["{{PrivateToken/operation}}"] is <code>"token-request"</code>:
+         1. If [$Should request be allowed to use feature?$] for "<code>[=policy-controlled feature/private-state-token-issuance=]</code>" returns false, then throw  "{{NotAllowedError}}" {{DOMException}} and abort the remaining steps.
```
Thanks. Fixed. 

I was trying the other way around, link text first.

jyasskin commented on spec.bs:
```diff
@@ -439,13 +439,24 @@ A [=/request=] has an associated <dfn for=request>private token issuers</dfn>, w
-     1. If |init|["{{RequestInit/privateToken}}"]["{{PrivateToken/operation}}"] is <code>"token-request"</code>, then abort these steps.
+     1. If |init|["{{RequestInit/privateToken}}"]["{{PrivateToken/operation}}"] is <code>"token-request"</code>:
+         1. If [$Should request be allowed to use feature?|should request be allowed$] for "<code>[=policy-controlled feature/private-state-token-issuance=]</code>" returns false, then throw a "{{NotAllowedError}}" {{DOMException}}.
+         1. Abort the remaining steps.
```
This will skip the rest of the algorithm for any call with init.privateToken.operation=="token-request", not only if permission policy disallows the call.

Strictly, you don't even need to say to abort the steps after you throw: https://infra.spec.whatwg.org/#algorithm-control-flow and https://webidl.spec.whatwg.org/#es-handling-exceptions say that throwing implies skipping the remaining steps. So you can just delete this line to fix the problem.

aykutbulut replied:
Algorithm is done if operation is `"token-request"`. Skipping rest of the algorithm is the intended behavior.

jyasskin replied:
Since this is a monkeypatch to the `new Request` algorithm in Fetch, aborting here means that it'll skip the rest of that constructor ... and not return anything as the new Request object. If the control flow you want is to skip the rest of the contents of the "If" block above, the clearest way to do that is probably to extract the body into a named algorithm. That'll also make merging the monkeypatch easier.

jyasskin commented on spec.bs:
```diff
@@ -439,13 +439,24 @@ A [=/request=] has an associated <dfn for=request>private token issuers</dfn>, w
+     1. If |init|["{{RequestInit/privateToken}}"]["{{PrivateToken/operation}}"] is <code>"token-request"</code>:
+         1. If [$Should request be allowed to use feature?|should request be allowed$] for "<code>[=policy-controlled feature/private-state-token-issuance=]</code>" returns false, then throw a "{{NotAllowedError}}" {{DOMException}}.
+         1. Abort the remaining steps.
+     1. If [$Should request be allowed to use feature?|should request be allowed$] for "<code>[=policy-controlled feature/private-state-token-redemption=]</code>" returns false:
```
This will check that `"private-state-token-redemption"` is allowed even for calls with init.privateToken.operation=="token-request". f4350027ad7858c7abe0fd9affc5e9e8f1d28d42 indented the 'then' steps, but I don't see any changes to the logic.

aykutbulut replied:
This part will not get executed if we keep the first `Abort the remaining steps.`. 

This `Abort the remaining steps.` is redundant and I removed it.

jyasskin commented on spec.bs:
```diff
@@ -439,13 +439,24 @@ A [=/request=] has an associated <dfn for=request>private token issuers</dfn>, w
      1. Set <var ignore>request</var>'s [=request/private token operation=]</a> to |init|["{{RequestInit/privateToken}}"]["{{PrivateToken/operation}}"].
-     1. If |init|["{{RequestInit/privateToken}}"]["{{PrivateToken/operation}}"] is <code>"token-request"</code>, then abort these steps.
+     1. If |init|["{{RequestInit/privateToken}}"]["{{PrivateToken/operation}}"] is <code>"token-request"</code>:
+         1. If [$Should request be allowed to use feature?|should request be allowed$] for "<code>[=policy-controlled feature/private-state-token-issuance=]</code>" returns false, then throw a "{{NotAllowedError}}" {{DOMException}}.
```
Oops, and this algorithm also takes the |request| as an argument, so you should pass that in both calls.

aykutbulut replied:
I see. I think using (1) might be more suitable here. We can not use (2) because request is being constructed. I wanted to use (2) since it returns `true` or `false`. (1) returns `"Enabled"` or `"Disabled"`. 

(1) https://w3c.github.io/webappsec-permissions-policy/#is-feature-enabled
(2) https://w3c.github.io/webappsec-permissions-policy/#should-request-be-allowed-to-use-feature

I will push the change soon.

aykutbulut replied:
Created a new algorithm and moved request modification steps to there.

dvorak42 commented on spec.bs:
```diff
@@ -374,6 +374,31 @@ To <dfn>unmask tokens</dfn> given [=key commitments=] |issuerKeys|, byte string
 1. Return |result|.
 
+To <dfn>set private token property for request</dfn> given a {{RequestInfo}} |input|, a {{RequestInit}} |init| and a {{Request}} |request| run the following steps:
+1. If |init|["{{RequestInit/privateToken}}"] [=map/exists=]:
```
nit: I wonder if you should move this check to the monkeypatch so that you only call this when there is a privateToken field to check?

aykutbulut replied:
Yes, I think this improves a lot. Prevents the large if block the algorithm had. Fixed.

jyasskin commented on spec.bs:
```diff
@@ -374,6 +374,30 @@ To <dfn>unmask tokens</dfn> given [=key commitments=] |issuerKeys|, byte string
 1. Return |result|.
 
+To <dfn>set private token property for request</dfn> given a {{RequestInfo}} |input|, a {{RequestInit}} |init| and a {{Request}} |request| run the following steps:
+1. Set |window| to |request|’s [=request/window=].
+1. Set |document| to |window|’s [=associated Document=].
+1. Let |origin| be |request|’s origin.
```
Since these are all introducing variables, they should all be "Let": https://infra.spec.whatwg.org/#variables

jyasskin commented on spec.bs:
```diff
@@ -374,6 +374,30 @@ To <dfn>unmask tokens</dfn> given [=key commitments=] |issuerKeys|, byte string
+1. If |init|["{{RequestInit/privateToken}}"]["{{PrivateToken/operation}}"] is <code>"token-request"</code>:
+    1. If [$Is feature enabled in document for origin?|is feature enabled$] on "<code>[=policy-controlled feature/private-state-token-issuance=]</code>", |document| and |origin| returns `"Disabled"`, then throw a "{{NotAllowedError}}" {{DOMException}}.
+    1. Abort the remaining steps.
+1. If [$Is feature enabled in document for origin?|is feature enabled$] on "<code>[=policy-controlled feature/private-state-token-redemption=]</code>", |document| and |origin| returns `"Disabled"`, then throw a "{{NotAllowedError}}" {{DOMException}}.
```
This could be a good place for a
```suggestion
1. Assert: |init|["{{RequestInit/privateToken}}"]["{{PrivateToken/operation}}"] is {{OperationType/"token-redemption"}} or {{OperationType/"send-redemption-record"}}
1. If [$Is feature enabled in document for origin?|is feature enabled$] on "<code>[=policy-controlled feature/private-state-token-redemption=]</code>", |document| and |origin| returns `"Disabled"`, then throw a "{{NotAllowedError}}" {{DOMException}}.
```

That could replace the assertion that it's "send-redemption-record" below, since it's easy to see the algorithm excluding "token-redemption" once you have a list of the possible values.

jyasskin commented on spec.bs:
```diff
@@ -374,6 +374,30 @@ To <dfn>unmask tokens</dfn> given [=key commitments=] |issuerKeys|, byte string
+1. Set |document| to |window|’s [=associated Document=].
+1. Let |origin| be |request|’s origin.
+1. Set |request|'s [=request/private token operation=]</a> to |init|["{{RequestInit/privateToken}}"]["{{PrivateToken/operation}}"].
+1. If |init|["{{RequestInit/privateToken}}"]["{{PrivateToken/operation}}"] is <code>"token-request"</code>:
```
You should link the enum value to its definition with
```suggestion
1. If |init|["{{RequestInit/privateToken}}"]["{{PrivateToken/operation}}"] is {{OperationType/"token-request"}}:
```

jyasskin commented on spec.bs:
```diff
@@ -374,6 +374,30 @@ To <dfn>unmask tokens</dfn> given [=key commitments=] |issuerKeys|, byte string
+    1. Abort the remaining steps.
+1. If [$Is feature enabled in document for origin?|is feature enabled$] on "<code>[=policy-controlled feature/private-state-token-redemption=]</code>", |document| and |origin| returns `"Disabled"`, then throw a "{{NotAllowedError}}" {{DOMException}}.
+1. If |init|["{{RequestInit/privateToken}}"]["{{PrivateToken/operation}}"] is <code>"token-redemption"</code>:
+    1. Set <var ignore>request</var>'s [=request/private token refresh policy=]</a> to |init|["{{RequestInit/privateToken}}"]["{{PrivateToken/refreshPolicy}}"].
```
You can go back to the `||` shorthand now that you have other uses of `|request|`.
```suggestion
    1. Set |request|'s [=request/private token refresh policy=]</a> to |init|["{{RequestInit/privateToken}}"]["{{PrivateToken/refreshPolicy}}"].
```
